File: src/main/java/com/fengsheng/card/Deck.java
package com.fengsheng.card;

import com.fengsheng.Game;
import com.fengsheng.HumanPlayer;
import com.fengsheng.Player;
import com.fengsheng.protos.Common;
import com.fengsheng.protos.Fengsheng;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class Deck {
    private int nextId;
    private final Game game;
    private final List<Card> cards = new ArrayList<>();
    private final List<Card> discardPile = new ArrayList<>();

    public Deck(Game game, boolean removeShiTan) {
        this.game = game;
        cards.addAll(DefaultDeck);
        if (removeShiTan)
            cards.subList(0, 18).clear();
        this.nextId = DefaultDeck.size();
        Collections.shuffle(cards);
    }

    /**
     * 洗牌
     */
    public void shuffle() {
        cards.addAll(discardPile);
        discardPile.clear();
        Collections.shuffle(cards);
        notifyDeckCount(true);
    }

    /**
     * 摸牌
     */
    public Card[] draw(int n) {
        if (n > cards.size()) {
            shuffle();
        }
        if (n > cards.size()) {
            n = cards.size();
        }
        List<Card> subList = cards.subList(cards.size() - n, cards.size());
        Card[] result = subList.toArray(new Card[0]);
        subList.clear();
        notifyDeckCount(false);
        if (cards.isEmpty()) {
            shuffle();
        }
        return result;
    }

    /**
     * 查看牌堆顶的n张牌
     *
     * @return 返回查看的牌，它是牌堆的一个 {@code sublist} ，其中 {@code cards.get(0)} 是观看的最后一张牌，
     * {@code cards.get(cards.length() - 1)} 是牌堆顶第一张牌。
     */
    public List<Card> peek(int n) {
        if (n > cards.size()) {
            shuffle();
        }
        if (n > cards.size()) {
            n = cards.size();
        }
        return cards.subList(cards.size() - n, cards.size());
    }

    /**
     * 弃牌
     */
    public void discard(Card... cards) {
        discardPile.addAll(List.of(cards));
    }

    /**
     * 获取牌堆剩余卡牌数量
     */
    public int getDeckCount() {
        return cards.size();
    }

    /**
     * 通知客户端牌堆剩余卡牌数量
     *
     * @param shuffled 是否洗牌
     */
    private void notifyDeckCount(boolean shuffled) {
        for (Player player : game.getPlayers()) {
            if (player instanceof HumanPlayer p) {
                p.send(Fengsheng.sync_deck_num_toc.newBuilder().setNum(cards.size()).setShuffled(shuffled).build());
            }
        }
    }

    /**
     * 如果需要新增一张卡牌，调用这个函数可以新卡牌的卡牌ID
     */
    public int getNextId() {
        return ++nextId;
    }

    public static final List<Card> DefaultDeck = List.of(
            new ShiTan(1, colors(Common.color.Black), Common.direction.Right, false, colors(Common.color.Black)),
            new ShiTan(2, colors(Common.color.Black), Common.direction.Right, false, colors(Common.color.Blue)),
            new ShiTan(3, colors(Common.color.Black), Common.direction.Right, false, colors(Common.color.Red, Common.color.Black)),
            new ShiTan(4, colors(Common.color.Black), Common.direction.Left, false, colors(Common.color.Red, Common.color.Blue)),
            new ShiTan(5, colors(Common.color.Black), Common.direction.Left, false, colors(Common.color.Blue, Common.color.Black)),
            new ShiTan(6, colors(Common.color.Black), Common.direction.Left, false, colors(Common.color.Red)),
            new ShiTan(7, colors(Common.color.Red), Common.direction.Right, false, colors(Common.color.Black)),
            new ShiTan(8, colors(Common.color.Red), Common.direction.Right, false, colors(Common.color.Blue)),
            new ShiTan(9, colors(Common.color.Red), Common.direction.Right, false, colors(Common.color.Red, Common.color.Black)),
            new ShiTan(10, colors(Common.color.Red), Common.direction.Left, false, colors(Common.color.Red, Common.color.Blue)),
            new ShiTan(11, colors(Common.color.Red), Common.direction.Left, false, colors(Common.color.Blue, Common.color.Black)),
            new ShiTan(12, colors(Common.color.Red), Common.direction.Left, false, colors(Common.color.Red)),
            new ShiTan(13, colors(Common.color.Blue), Common.direction.Right, false, colors(Common.color.Black)),
            new ShiTan(14, colors(Common.color.Blue), Common.direction.Right, false, colors(Common.color.Blue)),
            new ShiTan(15, colors(Common.color.Blue), Common.direction.Right, false, colors(Common.color.Red, Common.color.Black)),
            new ShiTan(16, colors(Common.color.Blue), Common.direction.Left, false, colors(Common.color.Red, Common.color.Blue)),
            new ShiTan(17, colors(Common.color.Blue), Common.direction.Left, false, colors(Common.color.Blue, Common.color.Black)),
            new ShiTan(18, colors(Common.color.Blue), Common.direction.Left, false, colors(Common.color.Red)),
            new PingHeng(19, colors(Common.color.Black), Common.direction.Left, true),
            new PingHeng(20, colors(Common.color.Black), Common.direction.Right, true),
            new PingHeng(21, colors(Common.color.Blue), Common.direction.Left, true),
            new PingHeng(22, colors(Common.color.Red), Common.direction.Right, true),
            new PingHeng(23, colors(Common.color.Red, Common.color.Black), Common.direction.Up, false),
            new PingHeng(24, colors(Common.color.Blue, Common.color.Black), Common.direction.Up, false),
            new PingHeng(25, colors(Common.color.Red, Common.color.Black), Common.direction.Left, false),
            new PingHeng(26, colors(Common.color.Blue, Common.color.Black), Common.direction.Right, false),
            new WeiBi(27, colors(Common.color.Red), Common.direction.Left, false),
            new WeiBi(28, colors(Common.color.Red), Common.direction.Left, false),
            new WeiBi(29, colors(Common.color.Red), Common.direction.Left, false),
            new WeiBi(30, colors(Common.color.Red), Common.direction.Right, false),
            new WeiBi(31, colors(Common.color.Blue), Common.direction.Left, false),
            new WeiBi(32, colors(Common.color.Blue), Common.direction.Right, false),
            new WeiBi(33, colors(Common.color.Blue), Common.direction.Right, false),
            new WeiBi(34, colors(Common.color.Blue), Common.direction.Right, false),
            new WeiBi(35, colors(Common.color.Black), Common.direction.Left, false),
            new WeiBi(36, colors(Common.color.Black), Common.direction.Left, false),
            new WeiBi(37, colors(Common.color.Black), Common.direction.Right, false),
            new WeiBi(38, colors(Common.color.Black), Common.direction.Right, false),
            new WeiBi(39, colors(Common.color.Blue, Common.color.Black), Common.direction.Left, false),
            new WeiBi(40, colors(Common.color.Red, Common.color.Black), Common.direction.Right, false),
            new LiYou(41, colors(Common.color.Black), Common.direction.Left, true),
            new LiYou(42, colors(Common.color.Black), Common.direction.Right, true),
            new LiYou(43, colors(Common.color.Black), Common.direction.Left, true),
            new LiYou(44, colors(Common.color.Black), Common.direction.Right, true),
            new LiYou(45, colors(Common.color.Black), Common.direction.Left, true),
            new LiYou(46, colors(Common.color.Black), Common.direction.Right, true),
            new LiYou(47, colors(Common.color.Blue), Common.direction.Left, true),
            new LiYou(48, colors(Common.color.Red), Common.direction.Right, true),
            new ChengQing(49, colors(Common.color.Red), Common.direction.Up, true),
            new ChengQing(50, colors(Common.color.Red), Common.direction.Up, true),
            new ChengQing(51, colors(Common.color.Black), Common.direction.Up, true),
            new ChengQing(52, colors(Common.color.Black), Common.direction.Up, true),
            new ChengQing(53, colors(Common.color.Blue), Common.direction.Up, true),
            new ChengQing(54, colors(Common.color.Blue), Common.direction.Up, true),
            new ChengQing(55, colors(Common.color.Black), Common.direction.Up, true),
            new ChengQing(56, colors(Common.color.Black), Common.direction.Up, true),
            new PoYi(57, colors(Common.color.Red, Common.color.Black), Common.direction.Left, true),
            new PoYi(58, colors(Common.color.Blue, Common.color.Black), Common.direction.Left, true),
            new PoYi(59, colors(Common.color.Red), Common.direction.Left, true),
            new PoYi(60, colors(Common.color.Blue), Common.direction.Left, true),
            new PoYi(61, colors(Common.color.Black), Common.direction.Left, true),
            new PoYi(62, colors(Common.color.Red, Common.color.Black), Common.direction.Right, true),
            new PoYi(63, colors(Common.color.Blue, Common.color.Black), Common.direction.Right, true),
            new PoYi(64, colors(Common.color.Red), Common.direction.Right, true),
            new PoYi(65, colors(Common.color.Blue), Common.direction.Right, true),
            new PoYi(66, colors(Common.color.Black), Common.direction.Right, true),
            new DiaoBao(67, colors(Common.color.Red), Common.direction.Up, false),
            new DiaoBao(68, colors(Common.color.Red), Common.direction.Left, false),
            new DiaoBao(69, colors(Common.color.Red), Common.direction.Right, false),
            new DiaoBao(70, colors(Common.color.Blue), Common.direction.Up, false),
            new DiaoBao(71, colors(Common.color.Blue), Common.direction.Left, false),
            new DiaoBao(72, colors(Common.color.Blue), Common.direction.Right, false),
            new DiaoBao(73, colors(Common.color.Black), Common.direction.Left, false),
            new DiaoBao(74, colors(Common.color.Black), Common.direction.Right, false),
            new DiaoBao(75, colors(Common.color.Red, Common.color.Black), Common.direction.Up, false),
            new DiaoBao(76, colors(Common.color.Red, Common.color.Black), Common.direction.Right, false),
            new DiaoBao(77, colors(Common.color.Blue, Common.color.Black), Common.direction.Up, false),
            new DiaoBao(78, colors(Common.color.Blue, Common.color.Black), Common.direction.Left, false),
            new JieHuo(79, colors(Common.color.Red), Common.direction.Up, false),
            new JieHuo(80, colors(Common.color.Red), Common.direction.Up, false),
            new JieHuo(81, colors(Common.color.Red), Common.direction.Up, true),
            new JieHuo(82, colors(Common.color.Blue), Common.direction.Up, false),
            new JieHuo(83, colors(Common.color.Blue), Common.direction.Up, false),
            new JieHuo(84, colors(Common.color.Blue), Common.direction.Up, true),
            new JieHuo(85, colors(Common.color.Black), Common.direction.Up, false),
            new JieHuo(86, colors(Common.color.Black), Common.direction.Up, false),
            new JieHuo(87, colors(Common.color.Black), Common.direction.Up, true),
            new JieHuo(88, colors(Common.color.Black), Common.direction.Up, true),
            new JieHuo(89, colors(Common.color.Blue, Common.color.Black), Common.direction.Up, false),
            new JieHuo(90, colors(Common.color.Red, Common.color.Black), Common.direction.Up, false),
            new WuDao(91, colors(Common.color.Red), Common.direction.Up, false),
            new WuDao(92, colors(Common.color.Red), Common.direction.Left, false),
            new WuDao(93, colors(Common.color.Red), Common.direction.Right, false),
            new WuDao(94, colors(Common.color.Blue), Common.direction.Up, false),
            new WuDao(95, colors(Common.color.Blue), Common.direction.Left, false),
            new WuDao(96, colors(Common.color.Blue), Common.direction.Right, false),
            new WuDao(97, colors(Common.color.Black), Common.direction.Left, false),
            new WuDao(98, colors(Common.color.Black), Common.direction.Right, false),
            new WuDao(99, colors(Common.color.Blue, Common.color.Black), Common.direction.Left, false),
            new WuDao(100, colors(Common.color.Red, Common.color.Black), Common.direction.Right, false)
    );

    private static Common.color[] colors(Common.color... colors) {
        return colors;
    }
}

File: src/main/java/card/Deck.kt
package com.fengsheng.card

import com.fengsheng.Game
import com.fengsheng.HumanPlayer
import com.fengsheng.protos.Common.color
import com.fengsheng.protos.Common.direction
import com.fengsheng.protos.Fengsheng

class Deck(private val game: Game) {
    private var nextId = 0
    private var cards = ArrayList<Card>()
    private var discardPile = ArrayList<Card>()

    /**
     * 洗牌
     */
    fun shuffle() {
        discardPile.shuffle()
        discardPile.addAll(cards)
        cards = discardPile
        discardPile = ArrayList()
        notifyDeckCount(true)
    }

    /**
     * 摸牌
     */
    fun draw(n: Int): Array<Card> {
        if (n > cards.size) shuffle()
        val from = if (n > cards.size) 0 else cards.size - n
        val subList = cards.subList(from, cards.size)
        val result = subList.toTypedArray()
        subList.clear()
        notifyDeckCount(false)
        if (cards.isEmpty()) {
            shuffle()
        }
        return result
    }

    /**
     * 查看牌堆顶的n张牌
     *
     * @return 返回查看的牌，它是牌堆的一个 `sublist` ，其中 `cards.get(0)` 是观看的最后一张牌，
     * `cards.get(cards.length() - 1)` 是牌堆顶第一张牌。
     */
    fun peek(n: Int): MutableList<Card> {
        if (n > cards.size) shuffle()
        val from = if (n > cards.size) 0 else cards.size - n
        return cards.subList(from, cards.size)
    }

    /**
     * 弃牌
     */
    fun discard(vararg cards: Card) {
        discardPile.addAll(cards)
    }

    val deckCount: Int
        /**
         * 获取牌堆剩余卡牌数量
         */
        get() = cards.size

    /**
     * 通知客户端牌堆剩余卡牌数量
     *
     * @param shuffled 是否洗牌
     */
    private fun notifyDeckCount(shuffled: Boolean) {
        for (player in game.players) {
            (player as? HumanPlayer)?.send(
                Fengsheng.sync_deck_num_toc.newBuilder().setNum(cards.size).setShuffled(shuffled).build()
            )
        }
    }

    /**
     * 如果需要新增一张卡牌，调用这个函数可以新卡牌的卡牌ID
     */
    fun getNextId(): Int {
        return ++nextId
    }

    fun init(totalPlayerCount: Int) {
        cards.clear()
        cards.addAll(DefaultDeck)
        if (totalPlayerCount < 4) {
            cards.subList(0, 18).clear()
        } else if (totalPlayerCount == 4 || totalPlayerCount == 5) {
            cards.removeAt(16)
            cards.removeAt(13)
            cards.removeAt(11)
            cards.removeAt(8)
            cards.removeAt(3)
            cards.removeAt(0)
        }
        nextId = DefaultDeck.size
        cards.shuffle()
    }

    companion object {
        val DefaultDeck = listOf(
            ShiTan(1, arrayOf(color.Black), direction.Right, false, arrayOf(color.Black)),
            ShiTan(2, arrayOf(color.Black), direction.Right, false, arrayOf(color.Blue)),
            ShiTan(3, arrayOf(color.Black), direction.Right, false, arrayOf(color.Red, color.Black)),
            ShiTan(4, arrayOf(color.Black), direction.Left, false, arrayOf(color.Red, color.Blue)),
            ShiTan(5, arrayOf(color.Black), direction.Left, false, arrayOf(color.Blue, color.Black)),
            ShiTan(6, arrayOf(color.Black), direction.Left, false, arrayOf(color.Red)),
            ShiTan(7, arrayOf(color.Red), direction.Right, false, arrayOf(color.Black)),
            ShiTan(8, arrayOf(color.Red), direction.Right, false, arrayOf(color.Blue)),
            ShiTan(9, arrayOf(color.Red), direction.Right, false, arrayOf(color.Red, color.Black)),
            ShiTan(10, arrayOf(color.Red), direction.Left, false, arrayOf(color.Red, color.Blue)),
            ShiTan(11, arrayOf(color.Red), direction.Left, false, arrayOf(color.Blue, color.Black)),
            ShiTan(12, arrayOf(color.Red), direction.Left, false, arrayOf(color.Red)),
            ShiTan(13, arrayOf(color.Blue), direction.Right, false, arrayOf(color.Black)),
            ShiTan(14, arrayOf(color.Blue), direction.Right, false, arrayOf(color.Blue)),
            ShiTan(15, arrayOf(color.Blue), direction.Right, false, arrayOf(color.Red, color.Black)),
            ShiTan(16, arrayOf(color.Blue), direction.Left, false, arrayOf(color.Red, color.Blue)),
            ShiTan(17, arrayOf(color.Blue), direction.Left, false, arrayOf(color.Blue, color.Black)),
            ShiTan(18, arrayOf(color.Blue), direction.Left, false, arrayOf(color.Red)),
            PingHeng(19, arrayOf(color.Black), direction.Left, true),
            PingHeng(20, arrayOf(color.Black), direction.Right, true),
            PingHeng(21, arrayOf(color.Blue), direction.Left, true),
            PingHeng(22, arrayOf(color.Red), direction.Right, true),
            PingHeng(23, arrayOf(color.Red, color.Black), direction.Up, false),
            PingHeng(24, arrayOf(color.Blue, color.Black), direction.Up, false),
            PingHeng(25, arrayOf(color.Red, color.Black), direction.Left, false),
            PingHeng(26, arrayOf(color.Blue, color.Black), direction.Right, false),
            WeiBi(27, arrayOf(color.Red), direction.Left, false),
            WeiBi(28, arrayOf(color.Red), direction.Left, false),
            WeiBi(29, arrayOf(color.Red), direction.Left, false),
            WeiBi(30, arrayOf(color.Red), direction.Right, false),
            WeiBi(31, arrayOf(color.Blue), direction.Left, false),
            WeiBi(32, arrayOf(color.Blue), direction.Right, false),
            WeiBi(33, arrayOf(color.Blue), direction.Right, false),
            WeiBi(34, arrayOf(color.Blue), direction.Right, false),
            WeiBi(35, arrayOf(color.Black), direction.Left, false),
            WeiBi(36, arrayOf(color.Black), direction.Left, false),
            WeiBi(37, arrayOf(color.Black), direction.Right, false),
            WeiBi(38, arrayOf(color.Black), direction.Right, false),
            WeiBi(39, arrayOf(color.Blue, color.Black), direction.Left, false),
            WeiBi(40, arrayOf(color.Red, color.Black), direction.Right, false),
            LiYou(41, arrayOf(color.Black), direction.Left, true),
            LiYou(42, arrayOf(color.Black), direction.Right, true),
            LiYou(43, arrayOf(color.Black), direction.Left, true),
            LiYou(44, arrayOf(color.Black), direction.Right, true),
            LiYou(45, arrayOf(color.Black), direction.Left, true),
            LiYou(46, arrayOf(color.Black), direction.Right, true),
            LiYou(47, arrayOf(color.Blue), direction.Left, true),
            LiYou(48, arrayOf(color.Red), direction.Right, true),
            ChengQing(49, arrayOf(color.Red), direction.Up, true),
            ChengQing(50, arrayOf(color.Red), direction.Up, true),
            ChengQing(51, arrayOf(color.Black), direction.Up, true),
            ChengQing(52, arrayOf(color.Black), direction.Up, true),
            ChengQing(53, arrayOf(color.Blue), direction.Up, true),
            ChengQing(54, arrayOf(color.Blue), direction.Up, true),
            ChengQing(55, arrayOf(color.Black), direction.Up, true),
            ChengQing(56, arrayOf(color.Black), direction.Up, true),
            PoYi(57, arrayOf(color.Red, color.Black), direction.Left, true),
            PoYi(58, arrayOf(color.Blue, color.Black), direction.Left, true),
            PoYi(59, arrayOf(color.Red), direction.Left, true),
            PoYi(60, arrayOf(color.Blue), direction.Left, true),
            PoYi(61, arrayOf(color.Black), direction.Left, true),
            PoYi(62, arrayOf(color.Red, color.Black), direction.Right, true),
            PoYi(63, arrayOf(color.Blue, color.Black), direction.Right, true),
            PoYi(64, arrayOf(color.Red), direction.Right, true),
            PoYi(65, arrayOf(color.Blue), direction.Right, true),
            PoYi(66, arrayOf(color.Black), direction.Right, true),
            DiaoBao(67, arrayOf(color.Red), direction.Up, false),
            DiaoBao(68, arrayOf(color.Red), direction.Left, false),
            DiaoBao(69, arrayOf(color.Red), direction.Right, false),
            DiaoBao(70, arrayOf(color.Blue), direction.Up, false),
            DiaoBao(71, arrayOf(color.Blue), direction.Left, false),
            DiaoBao(72, arrayOf(color.Blue), direction.Right, false),
            DiaoBao(73, arrayOf(color.Black), direction.Left, false),
            DiaoBao(74, arrayOf(color.Black), direction.Right, false),
            DiaoBao(75, arrayOf(color.Red, color.Black), direction.Up, false),
            DiaoBao(76, arrayOf(color.Red, color.Black), direction.Right, false),
            DiaoBao(77, arrayOf(color.Blue, color.Black), direction.Up, false),
            DiaoBao(78, arrayOf(color.Blue, color.Black), direction.Left, false),
            JieHuo(79, arrayOf(color.Red), direction.Up, false),
            JieHuo(80, arrayOf(color.Red), direction.Up, false),
            JieHuo(81, arrayOf(color.Red), direction.Up, true),
            JieHuo(82, arrayOf(color.Blue), direction.Up, false),
            JieHuo(83, arrayOf(color.Blue), direction.Up, false),
            JieHuo(84, arrayOf(color.Blue), direction.Up, true),
            JieHuo(85, arrayOf(color.Black), direction.Up, false),
            JieHuo(86, arrayOf(color.Black), direction.Up, false),
            JieHuo(87, arrayOf(color.Black), direction.Up, true),
            JieHuo(88, arrayOf(color.Black), direction.Up, true),
            JieHuo(89, arrayOf(color.Blue, color.Black), direction.Up, false),
            JieHuo(90, arrayOf(color.Red, color.Black), direction.Up, false),
            WuDao(91, arrayOf(color.Red), direction.Up, false),
            WuDao(92, arrayOf(color.Red), direction.Left, false),
            WuDao(93, arrayOf(color.Red), direction.Right, false),
            WuDao(94, arrayOf(color.Blue), direction.Up, false),
            WuDao(95, arrayOf(color.Blue), direction.Left, false),
            WuDao(96, arrayOf(color.Blue), direction.Right, false),
            WuDao(97, arrayOf(color.Black), direction.Left, false),
            WuDao(98, arrayOf(color.Black), direction.Right, false),
            WuDao(99, arrayOf(color.Blue, color.Black), direction.Left, false),
            WuDao(100, arrayOf(color.Red, color.Black), direction.Right, false),
            FengYunBianHuan(101, arrayOf(color.Black), direction.Up, false),
            FengYunBianHuan(102, arrayOf(color.Black), direction.Up, false)
        )
    }
}